mindsdb/integrations/handlers/merlion_handler/adapters.py

```diff
@@ -0,0 +1,182 @@
+from merlion.models.anomaly.isolation_forest import IsolationForestConfig, IsolationForest
+from merlion.models.anomaly.windstats import WindStatsConfig, WindStats
+from merlion.models.automl.autoprophet import AutoProphetConfig, AutoProphet
+from merlion.models.automl.autosarima import AutoSarimaConfig, AutoSarima
+from merlion.models.forecast.smoother import MSESConfig, MSES
+from merlion.post_process.threshold import AggregateAlarms
+from merlion.transform.moving_average import DifferenceTransform
+from scipy.stats import norm
+
+from merlion.models.defaults import DefaultDetectorConfig, DefaultDetector, DefaultForecaster, DefaultForecasterConfig
+from merlion.utils import TimeSeries
+
+from enum import Enum
+import pandas as pd
+
+
+class MerlionArguments(Enum):
+    target_seq_index = "target_seq_index"
+    max_forecast_steps = "max_forecast_steps"
+    max_backstep = "max_backstep"
+    wind_sz = "wind_sz"
+    alm_threshold = "alm_threshold"
+    maxiter = "maxiter"
+
+
+class BaseMerlionForecastAdapter:
+    TARGET_SEQ_INDEX = 0
+    DEFAULT_MAX_FORECAST_STEPS = 100
+    DEFAULT_MAX_BACKSTEP = 60
+    DEFAULT_MAXITER = 5
+
+    def __init__(self, **kwargs):
+        self.max_forecast_steps = kwargs.get(MerlionArguments.max_forecast_steps.value, self.DEFAULT_MAX_FORECAST_STEPS)
+        self.max_backstep = kwargs.get(MerlionArguments.max_backstep.value, self.DEFAULT_MAX_BACKSTEP)
+        self.maxiter = kwargs.get(MerlionArguments.maxiter.value, self.DEFAULT_MAXITER)
+        self.model = None
+
+    def to_bytes(self):
+        return self.model.to_bytes()
+
+    def from_bytes(self, bytes):
+        self.model = self.model.from_bytes(bytes)
+
+    def to_train_dataframe(self, df: pd.DataFrame, target: str) -> pd.DataFrame:
+        columns = list(df.columns.values)
+        columns.remove(target)
+        columns.insert(self.TARGET_SEQ_INDEX, target)
+        return df[columns]
+
+    def train(self, df: pd.DataFrame, target: str):
+        df = self.to_train_dataframe(df=df, target=target)
+        train_data = TimeSeries.from_pd(df)
+        self.model.train(train_data=train_data)
+
+    def predict(self, df: pd.DataFrame, target: str) -> pd.DataFrame:
+        forecast_step = self.max_forecast_steps
+        df = df[df.index <= self.model.last_train_time + self.model.timedelta * forecast_step]
+        if len(list(df.columns.values)) == 0:
+            df[target] = 0
+        predict_data = TimeSeries.from_pd(df)
+        predict_pred, predict_err = self.model.forecast(time_stamps=predict_data.time_stamps)
+        return self.__prepare_forecast_return(target=target, pred_ts=predict_pred, err_ts=predict_err)
+
+    def __prepare_forecast_return(self, target: str, pred_ts: TimeSeries, err_ts: TimeSeries) -> pd.DataFrame:
+        pred_df: pd.DataFrame = pred_ts.to_pd()
+        if err_ts is None:
+            err_df = None
+        else:
+            err_df = pred_ts.to_pd()
```

Comment:
Why is `err_df` set to `pred_ts`? Seems a bit weird because it means joining a `df` with itself on line 77, right? Shouldn't it be `err_df = err_ts.to_pd()`?

Reply:
> Why is `err_df` set to `pred_ts`? Seems a bit weird because it means joining a `df` with itself on line 77, right? Shouldn't it be `err_df = err_ts.to_pd()`?

Here, I suppose sometimes the err_df will miss some lines, so I use the left join to keep the join result has the same lines as input.

Reply:
You are right, should be err_df = err_ts.to_pd()